### PR TITLE
[AutoSparkUT] Fix ORC reads with missing nested fields [databricks]

### LINKED ISSUE #15179: [BUG][AutoSparkUT]SchemaPruningSuite.testSchemaPruning failed on GPU
**Describe the bug**
[SchemaPruningSuite.testSchemaPruning](https://github.com/apache/spark/blob/v3.3.0/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala#L650) from OrcV1SchemaPruningSuite and OrcV2SchemaPruningSuite failed on GPU.

**Steps/Code to reproduce bug**
spark-shell command:
```
/data/spark-3.3.0-bin-hadoop3/bin/spark-shell   --master local[2]   --conf spark.sql.optimizer.excludedRules=org.apache.spark.sql.catalyst.optimizer.ConvertToLocalRelation,org.apache.spark.sql.catalyst.optimizer.ConstantFolding   --conf spark.rapids.sql.enabled=true   --conf spark.plugins=com.nvidia.spark.SQLPlugin   --conf spark.sql.queryExecutionListeners=org.apache.spark.sql.rapids.ExecutionPlanCaptureCallback   --conf spark.rapids.sql.explain=ALL   --conf spark.rapids.sql.test.isFoldableNonLitAllowed=true   --conf spark.rapids.sql.csv.read.decimal.enabled=true   --conf spark.rapids.sql.format.avro.enabled=true   --conf spark.rapids.sql.format.avro.read.enabled=true   --conf spark.rapids.sql.format.json.enabled=true   --conf spark.rapids.sql.format.json.read.enabled=true   --conf spark.rapids.sql.incompatibleDateFormats.enabled=true   --conf spark.rapids.sql.python.gpu.enabled=true   --conf spark.rapids.sql.rowBasedUDF.enabled=true   --conf spark.rapids.sql.window.collectList.enabled=true   --conf spark.rapids.sql.window.collectSet.enabled=true   --conf spark.rapids.sql.window.range.byte.enabled=true   --conf spark.rapids.sql.window.range.short.enabled=true   --conf spark.rapids.sql.expression.Ascii=true   --conf spark.rapids.sql.expression.Conv=true   --conf spark.rapids.sql.expression.GetJsonObject=true   --conf spark.rapids.sql.expression.JsonToStructs=true   --conf spark.rapids.sql.expression.StructsToJson=true   --conf spark.rapids.sql.exec.CollectLimitExec=true   --conf spark.rapids.sql.exec.FlatMapCoGroupsInPandasExec=true   --conf spark.rapids.sql.exec.WindowInPandasExec=true   --conf spark.rapids.sql.hasExtendedYearValues=false   --conf spark.unsafe.exceptionOnMemoryLeak=true   --conf spark.sql.session.timeZone=UTC --conf spark.driver.bindAddress=127.0.0.1 --conf spark.driver.host=127.0.0.1
```

Scala code:
```
  import java.nio.file.Files
  import org.apache.spark.sql.Row
  import org.apache.spark.sql.types._

  spark.conf.set("spark.sql.shuffle.partitions", "1")
  spark.conf.set("spark.sql.orc.enableVectorizedReader", "true")
  spark.conf.set("spark.sql.orc.enableNestedColumnVectorizedReader", "true")

  val base = Files.createTempDirectory("rapids_orc_missing_nested_field_").toString
  println(s"BASE = $base")

  val fullName = new StructType().add("first", StringType, nullable = true).add("middle", StringType, nullable = true).add("last", StringType, nullable = true)

  val fullSchema = new StructType().add("id", IntegerType, nullable = true).add("name", fullName, nullable = true).add("address", StringType, nullable = true).add("pets", IntegerType, nullable = true)

  val fullRows = Seq(
    Row(0, Row("Jane", "X.", "Doe"), "123 Main Street", 1),
    Row(1, Row("John", "Y.", "Doe"), "321 Wall Street", 3)
  )

  val briefName = new StructType().add("first", StringType, nullable = true).add("last", StringType, nullable = true)

  val briefSchema = new StructType().add("id", IntegerType, nullable = true).add("name", briefName, nullable = true).add("address", StringType, nullable = true)

  val briefRows = Seq(
    Row(2, Row("Janet", "Jones"), "567 Maple Drive"),
    Row(3, Row("Jim", "Jones"), "6242 Ash Street")
  )

  spark.conf.set("spark.rapids.sql.enabled", "false")

  spark.createDataFrame(sc.parallelize(fullRows), fullSchema).coalesce(1).write.format("orc").mode("overwrite").save(s"$base/contacts/p=1")

  spark.createDataFrame(sc.parallelize(briefRows), briefSchema).coalesce(1).write.format("orc").mode("overwrite").save(s"$base/contacts/p=2")

  val readSchema = new StructType().add("id", IntegerType, nullable = true).add("name", fullName, nullable = true).add("address", StringType, nullable = true).add("pets", IntegerType, nullable = true).add("p", IntegerType, nullable = true)

  def run(label: String, useV1: Boolean, gpu: Boolean): Unit = {
    spark.conf.set("spark.sql.sources.useV1SourceList", if (useV1) "orc" else "")
    spark.conf.set("spark.rapids.sql.enabled", gpu.toString)

    spark.read.format("orc").schema(readSchema).load(s"$base/contacts")
      .createOrReplaceTempView("contacts")

    val out = spark.sql("select name.middle from contacts where p = 2")

    println(s"==== $label PLAN ====")
    println(out.queryExecution.executedPlan)

    try {
      println(s"==== $label RESULT ====")
      println(out.collect().toSeq)
    } catch {
      case t: Throwable =>
        val root = Iterator.iterate(t)(_.getCause).takeWhile(_ != null).toSeq.last
        println(s"==== $label ERROR ====")
        println(s"${t.getClass.getName}: ${t.getMessage}")
        println(s"ROOT: ${root.getClass.getName}: ${root.getMessage}")
    }
  }

  run("V1 CPU", useV1 = true, gpu = false)
  run("V1 GPU", useV1 = true, gpu = true)
  run("V2 CPU", useV1 = false, gpu = false)
  run("V2 GPU", useV1 = false, gpu = true)
```

output:
```
==== V1 CPU PLAN ====
*(1) Project [name#57.middle AS middle#66]
+- *(1) ColumnarToRow
   +- FileScan orc [name#57,p#60] Batched: true, DataFilters: [], Format: ORC, Location: InMemoryFileIndex(1 paths)[file:/tmp/rapids_orc_missing_nested_field_9001221213035628999/contacts], PartitionFilters: [isnotnull(p#60), (p#60 = 2)], PushedFilters: [], ReadSchema: struct<name:struct<middle:string>>

==== V1 CPU RESULT ====
WrappedArray([null], [null])

==== V1 GPU PLAN ====
26/07/01 20:06:35 WARN GpuOverrides: 
*Exec <ProjectExec> will run on GPU
  *Expression <Alias> name#76.middle AS middle#85 will run on GPU
    *Expression <GetStructField> name#76.middle will run on GPU
  *Exec <FileSourceScanExec> will run on GPU

GpuColumnarToRow false
+- GpuProject [name#76.middle AS middle#85], true
   +- GpuFileGpuScan orc [name#76] Batched: true, DataFilters: [], Eager_IO_Prefetch: false, Format: ORC, Location: InMemoryFileIndex[file:/tmp/rapids_orc_missing_nested_field_9001221213035628999/contacts], PartitionFilters: [isnotnull(p#79), (p#79 = 2)], PushedFilters: [], ReadSchema: struct<name:struct<middle:string>>

==== V1 GPU RESULT ====
26/07/01 20:06:36 ERROR Executor: Exception in task 0.0 in stage 5.0 (TID 5)
java.io.IOException: Error when processing path: file:///tmp/rapids_orc_missing_nested_field_9001221213035628999/contacts/p=2/part-00000-bcf3c470-05ba-4089-9dae-bdefec86131b-c000.snappy.orc, range: 0-766, partition values: [empty row]
	at com.nvidia.spark.rapids.OrcTableReader.$anonfun$next$1(GpuOrcScan.scala:2945)
	at com.nvidia.spark.rapids.NvtxId.apply(NvtxRangeWithDoc.scala:84)
	at com.nvidia.spark.rapids.NvtxIdWithMetrics$.apply(NvtxWithMetrics.scala:67)
	at com.nvidia.spark.rapids.OrcTableReader.next(GpuOrcScan.scala:2933)
	at com.nvidia.spark.rapids.OrcTableReader.next(GpuOrcScan.scala:2908)
	at com.nvidia.spark.rapids.CachedGpuBatchIterator$.$anonfun$apply$1(GpuDataProducer.scala:159)
	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
	at com.nvidia.spark.rapids.CachedGpuBatchIterator$.apply(GpuDataProducer.scala:156)
	at com.nvidia.spark.rapids.MultiFileCoalescingPartitionReaderBase.$anonfun$readBatchData$4(GpuMultiFileReader.scala:1369)
        ...

==== V2 CPU PLAN ====
*(1) Project [name#95.middle AS middle#104]
+- *(1) ColumnarToRow
   +- BatchScan orc file:/tmp/rapids_orc_missing_nested_field_9001221213035628999/contacts[name#95] OrcScan DataFilters: [], Format: orc, Location: InMemoryFileIndex(1 paths)[file:/tmp/rapids_orc_missing_nested_field_9001221213035628999/contacts], PartitionFilters: [isnotnull(p#98), (p#98 = 2)], PushedAggregation: [], PushedFilters: [], PushedGroupBy: [], ReadSchema: struct<name:struct<middle:string>> RuntimeFilters: []

==== V2 CPU RESULT ====
WrappedArray([null], [null])

==== V2 GPU PLAN ====
26/07/01 20:21:04 WARN GpuOverrides: 
*Exec <ProjectExec> will run on GPU
  *Expression <Alias> name#113.middle AS middle#122 will run on GPU
    *Expression <GetStructField> name#113.middle will run on GPU
  *Exec <BatchScanExec> will run on GPU
    *Input <OrcScan> will run on GPU

GpuColumnarToRow false
+- GpuProject [name#113.middle AS middle#122], true
   +- GpuBatchScan orc file:/tmp/rapids_orc_missing_nested_field_9001221213035628999/contacts[name#113] GpuOrcScan DataFilters: [], Format: gpuorc, Location: InMemoryFileIndex(1 paths)[file:/tmp/rapids_orc_missing_nested_field_9001221213035628999/contacts], PartitionFilters: [isnotnull(p#116), (p#116 = 2)], ReadSchema: struct<name:struct<middle:string>>, PushedFilters: [] RuntimeFilters: []

==== V2 GPU RESULT ====
26/07/01 20:21:04 ERROR Executor: Exception in task 0.0 in stage 7.0 (TID 7)
java.io.IOException: Error when processing path: file:///tmp/rapids_orc_missing_nested_field_9001221213035628999/contacts/p=2/part-00000-bcf3c470-05ba-4089-9dae-bdefec86131b-c000.snappy.orc, range: 0-766, partition values: []
	at com.nvidia.spark.rapids.OrcTableReader.$anonfun$next$1(GpuOrcScan.scala:2945)
	at com.nvidia.spark.rapids.NvtxId.apply(NvtxRangeWithDoc.scala:84)
	at com.nvidia.spark.rapids.NvtxIdWithMetrics$.apply(NvtxWithMetrics.scala:67)
	at com.nvidia.spark.rapids.OrcTableReader.next(GpuOrcScan.scala:2933)
	at com.nvidia.spark.rapids.OrcTableReader.next(GpuOrcScan.scala:2908)
	at com.nvidia.spark.rapids.CachedGpuBatchIterator$.$anonfun$apply$1(GpuDataProducer.scala:159)
	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
	at com.nvidia.spark.rapids.CachedGpuBatchIterator$.apply(GpuDataProducer.scala:156)
	at com.nvidia.spark.rapids.MultiFileCoalescingPartitionReaderBase.$anonfun$readBatchData$4(GpuMultiFileReader.scala:1369)
	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$AutoCloseableAttemptSpliterator.next(RmmRapidsRetryIterator.scala:581)
	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.next(RmmRapidsRetryIterator.scala:744)
        ...
```

**Expected behavior**
GPU should output as same as CPU does.

**Environment details (please complete the following information)**
 - Environment location: [local mode]

### PR DESCRIPTION
**Pull request by wjxiz1992**
**JaCoCo sql-plugin line coverage: +36 lines** (`GpuOrcScan.scala`; fix-line intersection, shim 330)

Closes #15179

## Problem

When every requested child of an ORC struct is absent from the physical file schema, nested pruning produced a zero-child struct. The GPU ORC reader then failed instead of returning the same nullable result as Spark CPU:

- V1 GPU: `ai.rapids.cudf.CudfException: std::bad_alloc`
- V2 GPU: `ai.rapids.cudf.CudfException: vector::_M_default_append`
- CPU: `WrappedArray([null], [null])`

## Fix

Retain the cheapest physical path under the struct when all requested children are missing. This preserves the parent validity stream so schema evolution can add the requested fields as null columns. Struct, list, and map paths keep their required ORC children and the include mask remains aligned with the physical schema.

The new Python integration coverage exercises V1 and V2 across all ORC reader configurations with primitive, array, and map physical children, including both non-null and null parent structs.

## Validation

- `mvn package -DskipTests -pl dist -am -Dbuildver=330 -Dmaven.repo.local=./.mvn-repo -s jenkins/settings.xml -P mirror-apache-to-urm`
  - `BUILD SUCCESS`
- `SPARK_HOME=/home/allxu/Downloads/spark-3.3.0-bin-hadoop3 TEST_PARALLEL=1 integration_tests/run_pyspark_from_build.sh -k 'test_read_orc_with_all_nested_fields_missing'`
  - `30 passed, 39164 deselected, 17 warnings in 29.50s`
- Full 32-config spark-shell reproduction after the fix:
  - V1 GPU and CPU: `WrappedArray([null], [null])`, match true
  - V2 GPU and CPU: `WrappedArray([null], [null])`, match true
- `scripts/check-shim-coverage.sh`: passed

## Performance impact

The schema traversal runs only when all requested children of a struct are absent, outside the row-processing hot path. It selects the shallowest and narrowest valid physical path to limit additional I/O. Normal matched-schema reads are unchanged.

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [ ] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [x] Not required

### CODE REVIEW

**Comment by wjxiz1992** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala [hunk 1]
```diff
@@ -1749,6 +1753,68 @@ private case class GpuOrcFileFilterHandler(
...
+      }
+
+      def retain(t: TypeDescription): TypeDescription = {
+        fileIncluded(t.getId) = true
+        t.getCategory match {
+          case TypeDescription.Category.STRUCT =>
+            val children = t.getChildren.asScala
+            val selectedIndex = children.indices.minBy(i => estimate(children(i)))
+            TypeDescription.createStruct().addField(
+              t.getFieldNames.get(selectedIndex), retain(children(selectedIndex)))
```
[the remark this answers is not in the document]
Comment (wjxiz1992) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala]: Addressed in e9dc09183. The STRUCT branch now returns a clone immediately when `children` is empty, avoiding `minBy` on an empty range while leaving normal schemas unchanged. Validation observed locally: the Spark 330 dist/integration reactor build succeeded, and the focused ORC GPU/CPU integration run passed all 30 cases.

**Review by res-life — COMMENTED** [review]
## Code Review: cudf-spark — PR #15210

**3 SHOULD FIX findings**

### 📄 `` `integration_tests/src/main/python/orc_test.py` `` ([diff](https://github.com/NVIDIA/cudf-spark/pull/15210/files#diff-443203b21132a865500ec72e1290faa84e199f96c120f1be9de4cf6019410519))

#### 1. [🟡 SHOULD FIX] Could this regression also cover an all-missing STRUCT inside a matching LIST element and a matching MAP value? At integration_tests/src/main/python/orc_test.py:638 and integration_tests/src/main/python/orc_test.py:647, `` `history` `` and `` `attrs` `` do not exist in the requested `` `name.middle` `` schema from integration_tests/src/main/python/orc_test.py:628

**Line:** 618 ([diff](https://github.com/NVIDIA/cudf-spark/pull/15210/files#diff-443203b21132a865500ec72e1290faa84e199f96c120f1be9de4cf6019410519R618))
**Issue:** Could this regression also cover an all-missing STRUCT inside a matching LIST element and a matching MAP value? At integration_tests/src/main/python/orc_test.py:638 and integration_tests/src/main/python/orc_test.py:647, `` `history` `` and `` `attrs` `` do not exist in the requested `` `name.middle` `` schema from integration_tests/src/main/python/orc_test.py:628. Execution therefore jumps from sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1721 into whole-path retention; it never exercises the changed recursive calls at sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1732 and sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1738 with an empty pruned child STRUCT. The pre-existing nested-array case at integration_tests/src/main/python/parquet_test.py:1055 keeps a matching child, so it does not cover this boundary either.
**Confidence:** 🟣❗ CERTAIN
**Code:**
```` ```python ````
```python
|
SELECT named_struct('history', array(1, 2)) AS name
...
lambda spark: spark.read.schema(read_schema).orc(array_data_path)
```
```` ``` ````
**Fix:** Add physical `` `array<struct<first:string>>` `` and `` `map<string,struct<first:string>>` `` inputs whose read schemas request `` `middle` `` inside the matching element/value STRUCT. These cases force the changed LIST and MAP recursion to reach the all-children-missing retention branch.
**Diff:**
```` ```diff ````
```diff
|
--- a/integration_tests/src/main/python/orc_test.py
+++ b/integration_tests/src/main/python/orc_test.py
@@
     assert_gpu_and_cpu_are_equal_collect(
         lambda spark: spark.read.schema(read_schema).orc(map_data_path), conf=all_confs)
+
+    nested_array_path = spark_tmp_path + '/ORC_NESTED_ARRAY_DATA'
+    with_cpu_session(
+        lambda spark: spark.sql("""
+            SELECT array(named_struct('first', 'Janet')) AS names
+            UNION ALL SELECT cast(null AS array<struct<first:string>>)
+            """).write.orc(nested_array_path))
+    nested_array_schema = StructType([StructField('names', ArrayType(StructType([
+        StructField('middle', StringType())])))])
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark: spark.read.schema(nested_array_schema).orc(nested_array_path),
+        conf=all_confs)
+
+    nested_map_path = spark_tmp_path + '/ORC_NESTED_MAP_DATA'
+    with_cpu_session(
+        lambda spark: spark.sql("""
+            SELECT map('primary', named_struct('first', 'Janet')) AS names
+            UNION ALL SELECT cast(null AS map<string,struct<first:string>>)
+            """).write.orc(nested_map_path))
+    nested_map_schema = StructType([StructField('names', MapType(
+        StringType(), StructType([StructField('middle', StringType())])))])
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark: spark.read.schema(nested_map_schema).orc(nested_map_path),
+        conf=all_confs)
```
```` ``` ````
**Reference:** integration_tests/src/main/python/cast_test.py:101 demonstrates nested `` `ArrayType` ``/StructType/MapType`` construction, and integration_tests/src/main/python/map_test.py:434 demonstrates a MAP containing a named STRUCT. The official Spark SQL reference confirms `` `named_struct` `` and `` `map` `` construct these values: https://spark.apache.org/docs/latest/sql-ref-functions-builtin.html
*(Reviewer: test-coverage)*

---

### 📄 `` `sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala` `` ([diff](https://github.com/NVIDIA/cudf-spark/pull/15210/files#diff-86813e200aa47eb04dcd8a7bf94491e0dc6d2fa67a09991a2641d947eaf52fc3))

#### 3. [🟡 SHOULD FIX] Can the regression matrix include an array-of-struct and map-to-struct whose requested struct fields are all absent? The changed recursion at sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1732, sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1738, and sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1741 must propagate `` `isRoot = false` `` so a STRUCT reached through LIST/MAP can retain a child

**Line:** 1732 ([diff](https://github.com/NVIDIA/cudf-spark/pull/15210/files#diff-86813e200aa47eb04dcd8a7bf94491e0dc6d2fa67a09991a2641d947eaf52fc3R1732))
**Issue:** Can the regression matrix include an array-of-struct and map-to-struct whose requested struct fields are all absent? The changed recursion at sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1732, sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1738, and sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1741 must propagate `` `isRoot = false` `` so a STRUCT reached through LIST/MAP can retain a child. The new array/map inputs at integration_tests/src/main/python/orc_test.py:638 and integration_tests/src/main/python/orc_test.py:647 instead put LIST/MAP below an already-unmatched outer STRUCT, so they exercise LIST/MAP in the retention helper, not these compatibility-recursion branches. Changing any of these three calls back to `` `isRoot = true` `` is not detected by the current tests.
**Confidence:** 🟣❗ CERTAIN
**Code:**
```` ```scala ````
```scala
|
case (TypeDescription.Category.LIST, TypeDescription.Category.LIST) =>
  val newChild = checkTypeCompatibility(..., isRoot = false)
case (TypeDescription.Category.MAP, TypeDescription.Category.MAP) =>
  val newKey = checkTypeCompatibility(..., isRoot = false)
  val newValue = checkTypeCompatibility(..., isRoot = false)
```
```` ``` ````
**Fix:** Extend integration_tests/src/main/python/orc_test.py:618 with physical `` `array<struct<first:string>>` `` and `` `map<string,struct<first:string>>` `` columns read as the same containers with `` `middle:string` ``. Keep the existing `` `assert_gpu_and_cpu_are_equal_collect` `` call at integration_tests/src/main/python/orc_test.py:632 so both V1/V2 and all reader configurations remain covered.
**Diff:**
```` ```diff ````
```diff
|
--- integration_tests/src/main/python/orc_test.py
+++ integration_tests/src/main/python/orc_test.py
@@
     assert_gpu_and_cpu_are_equal_collect(
         lambda spark: spark.read.schema(read_schema).orc(map_data_path), conf=all_confs)
+
+    array_struct_path = spark_tmp_path + '/ORC_ARRAY_STRUCT_DATA'
+    with_cpu_session(
+        lambda spark: spark.sql("""
+            SELECT array(named_struct('first', 'Janet'),
+                         cast(null AS struct<first:string>)) AS names
+            UNION ALL SELECT cast(null AS array<struct<first:string>>)
+            """).write.orc(array_struct_path))
+    array_struct_read_schema = StructType([StructField('names', ArrayType(StructType([
+        StructField('middle', StringType())])))])
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark: spark.read.schema(array_struct_read_schema).orc(array_struct_path),
+        conf=all_confs)
+
+    map_struct_path = spark_tmp_path + '/ORC_MAP_STRUCT_DATA'
+    with_cpu_session(
+        lambda spark: spark.sql("""
+            SELECT map('person', named_struct('first', 'Janet')) AS names
+            UNION ALL SELECT cast(null AS map<string,struct<first:string>>)
+            """).write.orc(map_struct_path))
+    map_struct_read_schema = StructType([StructField('names', MapType(
+        StringType(), StructType([StructField('middle', StringType())])))])
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark: spark.read.schema(map_struct_read_schema).orc(map_struct_path),
+        conf=all_confs)
```
```` ``` ````
**Reference:** integration_tests/src/main/python/parquet_test.py:1055 supplies the only inherited LIST-of-STRUCT pruning case, but it requests the existing `` `str_2` `` field and therefore does not enter the all-missing branch at sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1721. The project already constructs `` `ArrayType` ``/MapType`` nested schemas at integration_tests/src/main/python/cast_test.py:96 and integration_tests/src/main/python/cast_test.py:101.
*(Reviewer: test-coverage)*

#### 5. [🟡 SHOULD FIX] Can the empty-child STRUCT guard at sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1805 get a regression test? Removing this guard makes sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1808 call `` `minBy` `` on an empty range when the selected physical path ends in an empty STRUCT

**Line:** 1805 ([diff](https://github.com/NVIDIA/cudf-spark/pull/15210/files#diff-86813e200aa47eb04dcd8a7bf94491e0dc6d2fa67a09991a2641d947eaf52fc3R1805))
**Issue:** Can the empty-child STRUCT guard at sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1805 get a regression test? Removing this guard makes sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1808 call `` `minBy` `` on an empty range when the selected physical path ends in an empty STRUCT. The added cases at integration_tests/src/main/python/orc_test.py:623, integration_tests/src/main/python/orc_test.py:636, and integration_tests/src/main/python/orc_test.py:645 only create non-empty physical STRUCT children, so deleting the crash guard would leave every added assertion unchanged. This changed guard is a crash fix without a regression test.
**Confidence:** 🟣❗ CERTAIN
**Code:**
```` ```scala ````
```scala
|
case TypeDescription.Category.STRUCT =>
  val children = t.getChildren.asScala
  if (children.isEmpty) {
    t.clone()
  } else {
    val selectedIndex = children.indices.minBy(i => estimate(children(i)))
```
```` ``` ````
**Fix:** Add an ORC test resource under tests/src/test/resources whose physical schema contains a nested empty STRUCT, then read it with an evolved schema that requests an absent child. Include both a valid and a null parent row so the test verifies the guard and parent validity, and compare CPU and GPU results through the existing helper at tests/src/test/scala/com/nvidia/spark/rapids/SparkQueryCompareTestSuite.scala:2223.
**Diff:**
```` ```diff ````
```diff
|
--- a/tests/src/test/scala/com/nvidia/spark/rapids/OrcScanSuite.scala
+++ b/tests/src/test/scala/com/nvidia/spark/rapids/OrcScanSuite.scala
@@
+  // Add nested-empty-struct.orc to tests/src/test/resources with physical schema
+  // struct<name:struct<empty:struct<>>> and rows with valid/null `name` parents.
+  testSparkResultsAreEqual("schema evolution through an empty physical struct",
+    frameFromOrcWithSchema("nested-empty-struct.orc", StructType(Seq(
+      StructField("name", StructType(Seq(
+        StructField("missing", StringType)))))))) { frame => frame }
```
```` ``` ````
**Reference:** tests/src/test/scala/com/nvidia/spark/rapids/OrcScanSuite.scala:46 demonstrates the repository's ORC resource plus evolved-schema regression pattern; tests/src/test/scala/com/nvidia/spark/rapids/SparkQueryCompareTestSuite.scala:2223 resolves that resource and applies the requested schema. Apache ORC documents the STRUCT builder used to create the fixture: https://orc.apache.org/api/orc-core/org/apache/orc/TypeDescription.html
*(Reviewer: test-coverage)*

---

Full review generated locally; this comment contains only the SHOULD FIX findings requested for posting.

**Comment by wjxiz1992** [comment]
build

**Comment by wjxiz1992** [comment]
Added.

**Review by res-life — COMMENTED** [review]
NIT:
Only a test coverage issue remains; the production fix itself looks correct.

**SHOULD FIX — Make the regression test exercise the empty-child STRUCT guard**

In `tests/src/test/scala/com/nvidia/spark/rapids/OrcScanSuite.scala:57-58`, the physical `name` STRUCT contains both `empty:struct<>` and the cheaper `first:string` sibling. The cost selection in `sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala:1783-1808` therefore selects `first` and does not execute the empty-child guard at `GpuOrcScan.scala:1805-1806`. Removing that guard would still leave this test passing.

Please make `empty:struct<>` the only child of the physical `name` STRUCT and remove the now-unused STRING-vector setup. This is only a regression-test coverage issue, not a production-logic issue.

**Review by res-life — APPROVED** [review]
LGTM
